Fixes #27041

alpalla opened the pull request:
Fixes #27041. Added quick pick entry to terminate all running tasks along with argument 'terminateAll' which can be used with custom key binding to terminate all tasks via keyboard shortcut.
Possible issues/hacks associated with this fix:
- using difference between `undefined` and `null` to distinguish between "No task is currently running" and terminate "All running tasks" quick pick entries (https://github.com/alpalla/vscode/blob/master/src/vs/workbench/contrib/tasks/electron-browser/task.contribution.ts#L2219).
- using `showQuickPick()` parameter `selectedEntry` to actually add the terminate "All running tasks" entry to the quick pick (https://github.com/alpalla/vscode/blob/master/src/vs/workbench/contrib/tasks/electron-browser/task.contribution.ts#L1976).

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=72015) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=72015) before we can accept your contribution.<br/><br/>:x: alpalla [sign now](https://cla.opensource.microsoft.com/Microsoft/vscode?pullRequest=72015)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vscode?pullRequest=72015) it.</sub>

alexr00 commented on src/vs/workbench/contrib/tasks/electron-browser/task.contribution.ts:
```diff
@@ -1973,6 +1973,9 @@ class TaskService extends Disposable implements ITaskService {
 			if ((entries.length === 0) && defaultEntry) {
 				entries.push(defaultEntry);
 			}
+			else if (selectedEntry) {
```
Rather than using the selectedEntry, let's add a new paramter to `showQuickPick` for additional `TaskQuickPickEntry` items.

alexr00 commented on src/vs/workbench/contrib/tasks/electron-browser/task.contribution.ts:
```diff
@@ -2195,16 +2198,27 @@ class TaskService extends Disposable implements ITaskService {
+				}
 			).then(task => {
 				if (task === undefined || task === null) {
+					if (task === null) {
```
Instead of having this null or undefined check, I'd prefer that `showQuickPick` be refactored to return the quick pick entry instead of the task. That way you can set an ID on the quick pick entry and check for it here.

alpalla commented:
@alexr00 As per your suggestions I did the following:
- added a parameter `additionalEntries?: TaskQuickPickEntry[]` to `showQuickPick()` so as to be able to pass in the terminate "All running tasks" entry as an argument (and make it possible to pass in further additional quick pick entries in the future).
- refactored `showQuickPick()` to return type `TaskQuickPickEntry` (can still return `undefined` or `null`). Because of this refactoring I did have to add code in other functions where a `Task` return type was still expected, the added code looks like this:
```
if (entry === undefined || entry === null) {
    return;
}
let task: Task | undefined | null = entry.task;
```
It makes for the same duplicate code in about 8 different locations in the file, let me know if I should implement some kind of `getTaskFromEntry()` function to clean that up a little.
- lastly in `runTerminateCommand()` I used an `id` value of `66` to check whether the terminate "All running tasks" entry was chosen.

alexr00 commented on src/vs/workbench/contrib/tasks/electron-browser/task.contribution.ts:
```diff
@@ -2195,16 +2210,33 @@ class TaskService extends Disposable implements ITaskService {
+				undefined,
+				[{
+					label: nls.localize('TaskService.terminateAllRunningTasks', 'All running tasks'),
+					id: '66',
```
To make it consistant, let's make the id `terminalAll`.

alexr00 commented on src/vs/workbench/contrib/tasks/electron-browser/task.contribution.ts:
```diff
@@ -2195,16 +2210,33 @@ class TaskService extends Disposable implements ITaskService {
-			).then(task => {
+				false, true,
+				undefined,
+				[{
```
Having the All running tasks option show in the quick pick when there is only one task is a bit confusing. We should only show it when there is > 1 task.

alpalla commented:
@alexr00 Confirming I made the following changes:
- changed the id to `terminateAll`
- collapsed the previously mentioned checks onto one line as you suggested
- only show the terminate "All running tasks" entry if there are more than one task by checking if `entries.length > 1` in `showQuickPick()` (if that's okay) like so: https://github.com/alpalla/vscode/blob/master/src/vs/workbench/contrib/tasks/electron-browser/task.contribution.ts#L1976

alexr00 commented:
@alpalla Thank you for adding this feature!
